### PR TITLE
feat(eslint-plugin): add strict-type-predicates

### PR DESCRIPTION
**Pull request by bradleyayers**
This is a direct port of the TSLint version, with as few changes as possible.

Notable differences:

- No warning emitted when `strictNullChecks` is not enabled.
- Shorter rule description
- Only a single set of tests (instead of split `unknown`, `strictNullChecks`, `noStrictNullChecks`)

### CODE REVIEW

**Comment by bradleyayers** [comment]
Hey @bradzacher any more thoughts on this?

**Comment by bradzacher** [comment]
It's on my todo list. I've had family visiting for a few weeks so my spare time has been sparse.

I'm about to start getting back into the review queue.

I've never done a local fork, sorry, so I don't have Amy advice for you. I'd assume it should be pretty simple because the publish step is literally just publish each package's directory.

**Comment by ark120202** [comment]
Looks like it has some intersection with `no-unnecessary-condition`, at least in that case:

```ts
const number = 1;
if (number === null) {}
if (number === undefined) {}
```

It doesn't handle that though:

```ts
const numberOrNull: number | null = 0;
if (numberOrNull != null) {}

const number: number = 0;
if (typeof number === 'string') {}
```

I wonder if there's really a need for multiple rules being so similar (conceptually)?
Would it maybe be better to improve `no-unnecessary-condition` instead?

**Comment by bradleyayers** [comment]
I like the idea of merging it into no-unnecessary-condition, that seems like it could be a good fit. I think the name is an improvement too. I feel like it might need some options to enable the new behaviour though, as it's quite strict and might be undesirable for some people.

**Comment by bradzacher** [comment]
I'm curious - how much value is there for you in these checks?
```ts
const numberOrNull: number | null = 0;
// Implicitly checks for `!== undefined`, which is always false.
if (numberOrNull != null) {
  return;
}
const numberOrUndefined: number | undefined = 0;
// Implicitly checks for `!== null`, which is always false.
if (numberOrNull != undefined) {
  return;
}
```

I've always held the opinion that null checks should always be made `[!=]= null`, because it catches the undefined case - treating `null` and `undefined` differently seems like a path to bugs IMO.

---

This does seem like it has value for obvious reasons:

```ts
const number: number = 0;
// Always false.
if (typeof number === 'string') {
  return;
}
```

**Comment by bradleyayers** [comment]
In general I've found the rule really valuable for killing stale code, increasing confidence in refactors, and readability (you know exactly what types are being dealt with). The trade-off of having more verbosity in exchange for more clarity has been one we've been happy with. 

Specifically regarding null/undefined comparisons, they're good for refactors where a null might be swapped for an undefined, but a `!== null` is left unchanged in the code. Like you say though, another option is to ban !== for null/undefined, and force null/undefined comparisons to use `!=`. I think that's a reasonable direction to go, and encourage consistency with optional chaining.

What rule do you use to enforce that?

**Comment by bradzacher** [comment]
> What rule do you use to enforce that?

The base eslint `eqeqeq` rule will cover it, if you pass the second argument:
`["error", "always", {"null": "never"}]` - [eslint demo](https://eslint.org/demo#eyJ0ZXh0IjoiLyogZXNsaW50IGVxZXFlcTogW1wiZXJyb3JcIiwgXCJhbHdheXNcIiwge1wibnVsbFwiOiBcIm5ldmVyXCJ9XSAqL1xuXG5pZiAoeCA9PSBudWxsKSB7fVxuaWYgKHggIT0gbnVsbCkge31cbmlmICh4ID09PSBudWxsKSB7fVxuaWYgKHggIT09IG51bGwpIHt9XG5cbmlmICh4ID09PSAxKSB7fVxuaWYgKHggIT09IDEpIHt9XG5pZiAoeCA9PSAxKSB7fVxuaWYgKHggIT0gMSkge31cbiIsIm9wdGlvbnMiOnsicGFyc2VyT3B0aW9ucyI6eyJlY21hVmVyc2lvbiI6OSwic291cmNlVHlwZSI6Im1vZHVsZSIsImVjbWFGZWF0dXJlcyI6eyJqc3giOnRydWV9fSwicnVsZXMiOnt9LCJlbnYiOnt9fX0=)

---

I'd be happy to have both in `no-unnecessary-condition`:
- the typeof check can probably be straight up built into the rule
- `== null` can be added as an option to the rule.
  - I think that it is ultimately a stylistic choice as to whether you use `x == null` or `x === null || x === undefined`, so I think having an option to be super strict and trust types completely is 100% okay!

File: packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts [hunk 1]
```diff
@@ -0,0 +1,626 @@
...
+}
+
+// typeof object
+typeof get<{}>() === "object";
+
+// === null / undefined
+// get<number | null>() === null;
+// get<number | undefined>() === undefined;
+// get<string | null | undefined>() == null;
+// get<string | null | undefined>() != undefined;
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts]: uncomment or delete

File: packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts [hunk 2]
```diff
@@ -0,0 +1,626 @@
...
+enum E {}
+typeof get<E>() === "number";`,
+      errors: [
+        {
+          messageId: 'expressionAlwaysTrue',
+          line: 4,
+          column: 1,
+        },
+      ],
+    },
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts]: for completeness, would be good to get a string enum test `enum E { a = 'aaa' }`.

File: packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts [hunk 3]
```diff
@@ -0,0 +1,626 @@
...
+      errors: [
+        {
+          messageId: 'expressionAlwaysFalse',
+          line: 3,
+          column: 1,
+        },
+      ],
+    },
+
+    // typeof symbol
```
Comment (bradzacher) [inline packages/eslint-plugin/tests/rules/strict-type-predicates.test.ts]: just cos you've got a theme going, would be good to add a test for `Symbol`.

**Comment by felixfbecker** [comment]
Would really love to see this, one of the last rules that still requires me to use TSLint!

**Comment by ark120202** [comment]
So how would this rule relate to `no-unnecessary-condition`, especially with #1659?

**Comment by bradzacher** [comment]
#1659 will deal with the null/undefined case.

But `no-unnecessary-condition` does not currently check the `typeof` operator.
No reason it couldn't be augmented with it. Just depends on how much functionality we want to load up into that one rule.

**Pull request closed without merging**